### ISSUE TITLE
Pushforward / differentiate through probability distribution

### ISSUE THREAD

**Issue by gdalle** [JuliaDecisionFocusedLearning/DifferentiableExpectations.jl#13]
Why was it needed in the first place? Given a differentiable probability distribution constructor $\theta \mapsto p(\theta)$, differentiate through $\theta \mapsto E[c(V)]$ where $V \sim p(\theta)$.

Not directly necessary for `Perturbed` because we can put $c \circ f$ in the Reinforce, but necessary for regularized because $c \circ f$ is not an LP (not amenable to FW).

```julia
struct Pushforward
    fixed_atoms_dist_constructor  # must return a FixedAtomsProbabilityDistribution
    post_processing
end

function (p::Pushforward)(theta)
    dist = p.fixed_atoms_dist_constructor(theta)
    return mean(p.post_processing, dist)
end
```

Actually it is necessary to define `compute_probability_distribution` here along with its `rrule` for Reinforce

https://github.com/JuliaDecisionFocusedLearning/InferOpt.jl/blob/59ab4fe21682e8c0a20bb55d4da3b4cb91f2cc0c/src/perturbed/abstract_perturbed.jl#L118-L153

The `Pushforward` struct is no longer necessary then

**Comment by gdalle** [comment]
This is essentially solved by #15, in which I found out that:

- We can derive formulas for the rrule of `empirical_distribution` in the Reinforce case, but it's more cumbersome in the Reparametrization case.
- In the Reinforce case, it's not obvious how to generalize baseline variance reduction to the distribution pullback, and I'm not sure it gives the same result as the complete rrule for `Reinforce` if we do.
- The [rrule we had in InferOpt.jl](https://github.com/JuliaDecisionFocusedLearning/InferOpt.jl/blob/59ab4fe21682e8c0a20bb55d4da3b4cb91f2cc0c/src/perturbed/abstract_perturbed.jl#L116-L153) for this seems deeply broken. For once, the baseline doesn't live in the correct space: it should be an average of `y` values, not an average of weights

**Comment by gdalle** [comment]
Solved by #19
